### PR TITLE
[Inductor] optimize transpose_mxn with bf16 data type

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117958

**Summary**
Add the vectorization implementation of `transpose_mxn` with BFloat16 data type when matrix size is 16X16 or 32X32 which observed in Stable Diffusion BF16.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_transpose_mxn_16_16_bf16_fp16
python -u -m pytest -s -v test_cpu_repro.py -k test_transpose_mxn_32_32_bf16_fp16
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by jgong5 — CHANGES_REQUESTED** [review]
How about avx256?

File: aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h [hunk 1]
```diff
@@ -936,6 +936,435 @@ Vectorized<BFloat16> inline fmadd(const Vectorized<BFloat16>& a,
...
+  // t[9]: 288 289 290 291 292 293 294 295 296 297 298 299 300 301 302 303 304 ... 319
+  // t[10]: 320 321 322 323 324 325 326 327 328 329 330 331 332 333 334 335 336 ... 351
+  // t[11]: 352 353 354 355 356 357 358 359 360 361 362 363 364 365 366 367 368 ... 383
+  // t[12]: 384 385 386 387 388 389 390 391 392 393 394 395 396 397 398 399 400 ... 415
+  // t[13]: 416 417 418 419 420 421 422 423 424 425 426 427 428 429 430 431 432 ... 447
+  // t[14]: 448 449 450 451 452 453 454 455 456 457 458 459 460 461 462 463 464 ... 479
+  // t[15]: 480 481 482 483 484 485 486 487 488 489 490 491 492 493 494 495 496 ... 511
+  // t[16]: 512 513 514 515 516 517 518 519 520 521 522 523 524 525 526 527 528 ... 543
+  // ...
+  // t[31]: 992 993 994 995 996 997 998 999 1000 1001 1002 1003 1004 1005 1006 1007 1008 ... 1023
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h:1313]: Thanks for the suggestion and removed.

File: aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h [hunk 2]
```diff
@@ -936,6 +936,435 @@ Vectorized<BFloat16> inline fmadd(const Vectorized<BFloat16>& a,
...
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h:1353]: Same to remove this part.

File: aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h [hunk 3]
```diff
@@ -936,6 +936,435 @@ Vectorized<BFloat16> inline fmadd(const Vectorized<BFloat16>& a,
...
+  // t[13]: 388 420 389 421 390 422 391 423 396 428 397 429 398 430 399 431 404 ... 447
+  // t[14]: 448 480 449 481 450 482 451 483 456 488 457 489 458 490 459 491 464 ... 507
+  // t[15]: 452 484 453 485 454 486 455 487 460 492 461 493 462 494 463 495 468 ... 511
+  // t[16]: 512 544 513 545 514 546 515 547 520 552 521 553 522 554 523 555 528 ... 571
+  // ...
+  // t[31]: 964 996 965 997 966 998 967 999 972 1004 973 1005 974 1006 975 1007 980 ... 1023
+  for (int i = 0; i < 16; ++i) {
+    d[i * 2] = _mm512_unpacklo_epi16(r[i * 2], r[i * 2 + 1]);
+    d[i * 2 + 1] = _mm512_unpackhi_epi16(r[i * 2], r[i * 2 + 1]);
+  }
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h:1141]: Thanks for the suggestion and added `#pragma unroll(n)` to force the compiler unroll these loops.

File: aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h [hunk 4]
```diff
@@ -936,6 +936,435 @@ Vectorized<BFloat16> inline fmadd(const Vectorized<BFloat16>& a,
...
+  // l0-l15  p0-p15
+  __m512i r[8];
+  r[0] = _mm512_inserti64x4(_mm512_castsi256_si512(t[0]), t[4], 0x01);
+  r[1] = _mm512_inserti64x4(_mm512_castsi256_si512(t[1]), t[5], 0x01);
+  r[2] = _mm512_inserti64x4(_mm512_castsi256_si512(t[2]), t[6], 0x01);
+  r[3] = _mm512_inserti64x4(_mm512_castsi256_si512(t[3]), t[7], 0x01);
+  r[4] = _mm512_inserti64x4(_mm512_castsi256_si512(t[8]), t[12], 0x01);
+  r[5] = _mm512_inserti64x4(_mm512_castsi256_si512(t[9]), t[13], 0x01);
+  r[6] = _mm512_inserti64x4(_mm512_castsi256_si512(t[10]), t[14], 0x01);
+  r[7] = _mm512_inserti64x4(_mm512_castsi256_si512(t[11]), t[15], 0x01);
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h:956]: Sure, simplify the code with loops.

**Comment by leslie-fang-intel** [comment]
> How about avx256?

Thanks for the suggestion. Add a TODO to enable the AVX2 counterparts. Can we enable it later?

**Review by jgong5 — APPROVED** [review]

**Review by lezcano — APPROVED** [review]
I have not reviewed, but @jgong5 has.

**Pull request closed without merging**